Mathlib/GroupTheory/Perm/MaximalSubgroups.lean

```diff
@@ -0,0 +1,430 @@
+  rw [Set.notMem_compl_iff] at ha'
+  obtain ⟨b, hb, hb'⟩ :=
+    Set.not_subset_iff_exists_mem_notMem.mp fun h => hB' ((hB_not_le_s B hB) h)
+  rw [← Set.mem_compl_iff] at hb'
+  have hsc_le_B : sᶜ ⊆ B := by
+    intro x hx'
+    suffices ∃ k : fixingSubgroup (Perm α) s, k • b = x by
+      obtain ⟨⟨k, hk⟩, hkbx : k • b = x⟩ := this
+      suffices k • B = B by
+        rw [← hkbx, ← this, Set.smul_mem_smul_set_iff]
+        exact hb
+      -- `k • B = B`
+      apply isBlock_iff_smul_eq_of_nonempty.mp hB (g := ⟨k, ?_⟩)
+      · refine ⟨a, ⟨?_, ha⟩⟩
+        rw [mem_fixingSubgroup_iff] at hk
+        rw [← hk a ha']
+        exact Set.smul_mem_smul_set ha
+      · -- `↑k ∈ G`
+        apply le_of_lt hG
+        exact MulAction.fixingSubgroup_le_stabilizer _ _ hk
+    · -- `∃ (k : fixingSubgroup (Perm α) s), k • b = x`
+      suffices
+        IsPretransitive (fixingSubgroup (Perm α) s)
+          (ofFixingSubgroup (Perm α) s) by
+        obtain ⟨k, hk⟩ :=
+          exists_smul_eq (fixingSubgroup (Perm α) s)
+           (⟨b, hb'⟩ : ofFixingSubgroup (Perm α) s) ⟨x, hx'⟩
+        use k
+        rw [← Subtype.coe_inj, val_smul] at hk
+        exact hk
+      -- Prove pretransitivity…
+      rw [← is_one_pretransitive_iff]
+      have _ :  IsMultiplyPretransitive (Perm α) α (s.ncard + 1) := by
+        exact isMultiplyPretransitive α (s.ncard + 1)
+      apply ofFixingSubgroup.isMultiplyPretransitive
+        (Perm α) s rfl
+  -- Conclusion of the proof : `B = univ`
+  rw [← Set.top_eq_univ, eq_top_iff]
+  intro x _
+  obtain ⟨b, hb⟩ := h1
+  obtain ⟨⟨g, hg⟩, hgbx : g • b = x⟩ := exists_smul_eq G b x
+  suffices g • B = B by
+    rw [← hgbx, ← this, Set.smul_mem_smul_set_iff]
+    exact hsc_le_B hb
+  -- `g • B = B`
+  apply isBlock_iff_smul_eq_of_nonempty.mp hB (g := ⟨g, hg⟩)
+  rw [Subgroup.mk_smul]
+  apply nonempty_inter_of_lt_ncard_add_ncard
+  -- `card B + card (g • B) = card B + card B`
+  -- ... ≥ `card sᶜ + card sᶜ`
+  -- ... > `card s + card s ᶜ = card α`
+  rw [← Set.ncard_add_ncard_compl s]
+  apply Nat.lt_of_lt_of_le
+  · rw [Nat.add_lt_add_iff_right]
+    exact hα
+  · simp only [Set.ncard_smul_set, ← two_mul]
+    apply Nat.mul_le_mul_left
+    exact Set.ncard_le_ncard hsc_le_B (Set.toFinite B)
+
+/-- `MulAction.stabilizer (Perm α) s` is a maximal subgroup of `Perm α`,
+provided `s` and `sᶜ` are nonempty, and `Nat.card α ≠ 2 * Nat.card s`. -/
+theorem isCoatom_stabilizer (s : Set α)
+    (h0 : s.Nonempty) (h1 : sᶜ.Nonempty)
+    (hα : Nat.card α ≠ 2 * s.ncard) :
+    IsCoatom (stabilizer (Perm α) s) := by
+  rcases Nat.lt_trichotomy s.ncard sᶜ.ncard with hs | hs'
+  · exact isCoatom_stabilizer_of_ncard_lt_ncard_compl s h0 h1 hs
+  rcases hs' with hs | hs
+  · exfalso; apply hα
+    rw [← Set.ncard_add_ncard_compl s, two_mul, ← hs]
+  · rw [← compl_compl s] at h0
+    rw [← stabilizer_compl]
+    apply isCoatom_stabilizer_of_ncard_lt_ncard_compl (sᶜ) h1 h0
```

Comment:
If you accept the removal of the `h1` hypothesis above, then you can do this here:
```suggestion
  obtain hs | hs | hs := Nat.lt_trichotomy s.ncard sᶜ.ncard
  · exact isCoatom_stabilizer_of_ncard_lt_ncard_compl s h0 hs
  · contrapose! hα
    rw [← Set.ncard_add_ncard_compl s, two_mul, ← hs]
  · rw [← stabilizer_compl]
    apply isCoatom_stabilizer_of_ncard_lt_ncard_compl sᶜ h1
```